test/e2e/cluster_upgrade.go

```diff
@@ -180,3 +211,90 @@ func ClusterUpgradeConformanceSpec(ctx context.Context, inputGetter func() Clust
 		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, clusterResources.Cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
 	})
 }
+
+// upgradeClusterTopologyAndWaitForUpgradeInput is the input type for upgradeClusterTopologyAndWaitForUpgrade.
+type upgradeClusterTopologyAndWaitForUpgradeInput struct {
+	ClusterProxy                framework.ClusterProxy
+	Cluster                     *clusterv1.Cluster
+	ControlPlane                *controlplanev1.KubeadmControlPlane
+	EtcdImageTag                string
+	DNSImageTag                 string
+	MachineDeployments          []*clusterv1.MachineDeployment
+	KubernetesUpgradeVersion    string
+	WaitForMachinesToBeUpgraded []interface{}
+	WaitForKubeProxyUpgrade     []interface{}
+	WaitForDNSUpgrade           []interface{}
+	WaitForEtcdUpgrade          []interface{}
+}
+
+// upgradeClusterTopologyAndWaitForUpgrade upgrades a Cluster topology and waits for it to be upgraded.
+func upgradeClusterTopologyAndWaitForUpgrade(ctx context.Context, input upgradeClusterTopologyAndWaitForUpgradeInput) {
```

Comment:
Sorry for the back and forth on this, but it seems to me that the resulting func in this current form is clean and we can move it into the framerwok give a disclaimer that it works with KCP only. WDYT?